From 24b55ff8f07093dbf056d8b9e1c60a32d8c4f772 Mon Sep 17 00:00:00 2001
From: "jets.yan" <jets.yan@amlogic.com>
Date: Thu, 12 Jun 2014 17:20:03 +0800
Subject: [PATCH 4331/5965] PD #91439: fix the display-flush occasionally
 during display-mode exchange for hdmi on one TCL tv.

---
 arch/arm/mach-meson8/hdmi_tx_hw/hdmi_tx_hw.c |  5 ++++-
 drivers/amlogic/display/vout/tvoutc.c        | 21 +++++++++++++++++++-
 2 files changed, 24 insertions(+), 2 deletions(-)

diff --git a/arch/arm/mach-meson8/hdmi_tx_hw/hdmi_tx_hw.c b/arch/arm/mach-meson8/hdmi_tx_hw/hdmi_tx_hw.c
index fa7e1ecf3619..46cc1ae1eba1 100755
--- a/arch/arm/mach-meson8/hdmi_tx_hw/hdmi_tx_hw.c
+++ b/arch/arm/mach-meson8/hdmi_tx_hw/hdmi_tx_hw.c
@@ -1939,7 +1939,8 @@ static int hdmitx_set_dispmode(hdmitx_dev_t* hdmitx_device, Hdmi_tx_video_para_t
     }
 
     hdmi_hw_reset(hdmitx_device, param);    
-    hdmitx_set_pll(param);
+	// move hdmitx_set_pll() to the end of this function.
+    // hdmitx_set_pll(param);
     hdmitx_set_phy(hdmitx_device);
 
     if((param->VIC==HDMI_720p60)||(param->VIC==HDMI_720p50)||
@@ -1987,6 +1988,8 @@ static int hdmitx_set_dispmode(hdmitx_dev_t* hdmitx_device, Hdmi_tx_video_para_t
     hdmi_wr_reg(TX_SYS5_TX_SOFT_RESET_2, 0x60);
     mdelay(5);
 
+	hdmitx_set_pll(param);
+
     return 0;
 }
 
diff --git a/drivers/amlogic/display/vout/tvoutc.c b/drivers/amlogic/display/vout/tvoutc.c
index 0f22ec6ebf0f..eae32d36b4d2 100755
--- a/drivers/amlogic/display/vout/tvoutc.c
+++ b/drivers/amlogic/display/vout/tvoutc.c
@@ -378,6 +378,18 @@ int tvoutc_setmode(tvmode_t mode)
     }
 
 #if ((defined CONFIG_ARCH_MESON8) || (defined CONFIG_ARCH_MESON8B))
+	// for hdmi mode, disable HPLL as soon as possible
+	if( (mode==TVMODE_480I) || (mode==TVMODE_480P) ||
+		(mode==TVMODE_576I) || (mode==TVMODE_576P) ||
+		(mode==TVMODE_720P) || (mode==TVMODE_720P_50HZ) ||
+		(mode==TVMODE_1080I) || (mode==TVMODE_1080I_50HZ) ||
+		(mode==TVMODE_1080P) || (mode==TVMODE_1080P_50HZ) ||
+		(mode==TVMODE_1080P_24HZ) || (mode==TVMODE_4K2K_24HZ) ||
+		(mode==TVMODE_4K2K_25HZ) || (mode==TVMODE_4K2K_30HZ) ||
+		(mode==TVMODE_4K2K_SMPTE) )
+	{
+		WRITE_CBUS_REG_BITS(HHI_VID_PLL_CNTL, 0x0, 30, 1);
+	}
     cvbs_cntl_output(0);
 #endif
     while (MREG_END_MARKER != s->reg)
@@ -390,7 +402,14 @@ int tvoutc_setmode(tvmode_t mode)
 	aml_write_reg32(P_PERIPHS_PIN_MUX_0,aml_read_reg32(P_PERIPHS_PIN_MUX_0)&(~(3<<20)));
     }
 
-    set_tvmode_misc(mode);
+#if ((defined CONFIG_ARCH_MESON8) || (defined CONFIG_ARCH_MESON8B))
+	// for hdmi mode, leave the hpll setting to be done by hdmi module.
+	if( (mode==TVMODE_480CVBS) || (mode==TVMODE_576CVBS) )
+		set_tvmode_misc(mode);
+#else
+	set_tvmode_misc(mode);
+#endif
+
 #ifdef CONFIG_ARCH_MESON1
 	tvoutc_setclk(mode);
     printk("%s[%d]\n", __func__, __LINE__);
-- 
2.19.0

